Switch to `process::Command` API instead of `duct`

TimJentzsch opened the pull request:
Closes #4, allows us to tackle #2.

This PR explores using `process::Command` instead of the `duct` library.

This allows us to remove a dependency while gaining more control about command execution.
In particular, it allow us to read the process output while still being able to kill the process, solving #4.

Note that the completed event doesn't always trigger immediately after killing the process, especially when coupled with the `sleep` command.
If using a command like `echo Sleeping for 10s && sleep 10 && echo Done`, the `sleep` is executed in a separate thread and not killed along with the process.
So the command will not send the "Done" output anymore, but still wait the remaining 10 seconds... not sure if we can fix that.

To enable this, the following changes were made:

- Remove `duct` and `futures-lite`, as they are not needed anymore.
- Store the `Command` and `Child` (running process) for each command.
- Only wrap the output buffer in an arc mutex and not the whole command info. This improves code ergonomics and probably also performance.
- Use a `HashMap` instead of `Vec` to store the running processes. This makes the lookup of a specific PID easier.
- Add new example to demonstrate killing a process. Note that the process still "lingers" because `sleep` is being run in a separate process.

TimJentzsch committed:
Use the process::Command API for fine grained control

This allows us to read the process output while still being able to kill the process

edouardpoitras commented:
FYI: The only "conflict" will be changing .iter() to .read() and using bevy 0.12.

TimJentzsch committed:
Only wrap output buffer in mutex

This simplifies the logic a lot and potentially improves performance

TimJentzsch committed:
Handle process completion

TimJentzsch committed:
Merge branch 'main' into use-command-api

TimJentzsch committed:
Ensure systems run in the correct order

TimJentzsch committed:
Fix output lines being duplicated

TimJentzsch committed:
Remove unused futures_lite dependency

TimJentzsch committed:
Add example of killing processes

TimJentzsch commented:
@edouardpoitras The good news is that the approach seems to work and that it allowed us to remove all dependencies except bevy.
I also think that this will make it quite easy to allow us writing to processes via events.

The bad news is that the PR got quite big, please let me know if it's too hard to review and I'll try to split it up.
I updated the PR description with an overview of the changes I made.

edouardpoitras commented on src/lib.rs:
```diff
@@ -72,234 +78,143 @@ impl Plugin for BevyLocalCommandsPlugin {
-                error!("Failed to add new active shell command: {}", &cs)
+fn spawn_process(mut command: Command) -> ActiveProcess {
+    // Start running the process
+    let mut process = command.spawn().unwrap();
```
I recognize there's still some hardening to be done, but for this PR can we handle the scenario where a invalid executable is provided instead of just panicking on the unwrap? I'm getting this output when program isn't valid:
```
thread 'Compute Task Pool (3)' panicked at src\lib.rs:184:39:
called `Result::unwrap()` on an `Err` value: Error { kind: NotFound, message: "program not found" }
```

Maybe spawn_process can return a Result instead? Open to other ideas.

TimJentzsch replied:
Good point, we should handle this.
Maybe having an event stream for errors where the user can react to this?

TimJentzsch replied:
@edouardpoitras Sorry for the delay, I now added an event for errors and use that when the process failed to start instead of crashing the app.
In future PRs we can expand this mechanism for other kinds of errors that can occur.

I also renamed most events from `command` to `process` in order to distinguish a bit from `std::process::Command`.
But if you don't like the new names I can also revert those changes.

edouardpoitras replied:
No worries, there's no rush here - I'm just happy to have some help.
No issues on the rename.
I added a few comments for review.
Thanks.

TimJentzsch committed:
Emit event when process failed to start

Also renames the events from command to process,
in order to distinguish from the std::process:Command

TimJentzsch committed:
Make variable and system names more consistent

edouardpoitras commented on src/lib.rs:
```diff
@@ -42,264 +46,186 @@ pub struct ShellCommandOutput {
+            .add_event::<ProcessStarted>()
+            .add_event::<ProcessOutputEvent>()
+            .add_event::<KillProcess>()
+            .add_event::<ProcessCompleted>()
```
Don't we need to add `ProcessError` as an event here?

TimJentzsch replied:
Ha, nice catch. I'll add that later today.

TimJentzsch replied:
Added

edouardpoitras commented on src/lib.rs:
```diff
@@ -42,264 +46,186 @@ pub struct ShellCommandOutput {
+            .stdout(Stdio::piped());
+
+        let Ok(active_process) = spawn_process(cmd) else {
+            process_error_event.send(ProcessError::FailedToStart);
```
Can we add the program/args to the event? Just so the user can reference which command failed to start.

TimJentzsch replied:
Good point.
For now, I added the command string, as the PID is not available yet.
Maybe in the future we can return something that's of more practical use.

edouardpoitras replied:
Perfect! Thanks @TimJentzsch

TimJentzsch committed:
Add basic command info to process error

TimJentzsch committed:
Register process error event in app
